server/controller/src/main/java/ai/starwhale/mlops/datastore/ColumnType.java

```diff
@@ -89,102 +101,168 @@ public static int compare(ColumnType type1, Object value1, ColumnType type2, Obj
+                    return Double.compare(((Number) value1).doubleValue(), ((Number) value2).doubleValue());
+                } else if (category1.equals(ColumnTypeScalar.FLOAT32.getCategory())
+                        && category2.equals(ColumnTypeScalar.INT32.getCategory())) {
+                    return -ColumnType.compare(type2, value2, type1, value1);
                 }
             }
-        } else if (type1 instanceof ColumnTypeList && type2 instanceof ColumnTypeList) {
-            var iter1 = ((List<?>) value1).iterator();
-            var iter2 = ((List<?>) value2).iterator();
-            for (; ; ) {
-                if (iter1.hasNext() && iter2.hasNext()) {
-                    var result = ColumnType.compare(((ColumnTypeList) type1).getElementType(), iter1.next(),
-                            ((ColumnTypeList) type2).getElementType(), iter2.next());
-                    if (result != 0) {
-                        return result;
-                    }
-                } else if (iter1.hasNext()) {
-                    return 1;
-                } else if (iter2.hasNext()) {
-                    return -1;
+        }
+        throw new IllegalArgumentException("can not compare " + type1 + " with " + type2);
+    }
+
+    private static int compareList(ColumnTypeList type1, Object value1, ColumnTypeList type2, Object value2) {
+        var iter1 = ((List<?>) value1).iterator();
+        var iter2 = ((List<?>) value2).iterator();
+        for (; ; ) {
+            if (iter1.hasNext() && iter2.hasNext()) {
+                var result = ColumnType.compare(type1.getElementType(),
+                        iter1.next(),
+                        type2.getElementType(),
+                        iter2.next());
+                if (result != 0) {
+                    return result;
+                }
+            } else if (iter1.hasNext()) {
+                return 1;
+            } else if (iter2.hasNext()) {
+                return -1;
+            } else {
+                return 0;
+            }
+        }
+    }
+
+    private static int compareMap(ColumnTypeMap type1, Object value1, ColumnTypeMap type2, Object value2) {
+        var map1 = (Map<?, ?>) value1;
+        var map2 = (Map<?, ?>) value2;
+        if (map1.isEmpty() && map2.isEmpty()) {
+            return 0;
+        }
+        if (map1.isEmpty()) {
+            return -1;
+        }
+        if (map2.isEmpty()) {
+            return 1;
+        }
+        Object minDiffKey = null;
+        int result = 0;
+        for (var iter = Stream.concat(map1.keySet().stream(), map2.keySet().stream()).iterator();
+                iter.hasNext(); ) {
+            var key = iter.next();
+            if (minDiffKey == null || ColumnType.compare(type1.getKeyType(), key, type1.getKeyType(), minDiffKey) < 0) {
+                var v1 = map1.get(key);
+                var v2 = map2.get(key);
+                int t;
+                if (v1 == null) {
+                    t = 1;
```

Comment:
-1?

Reply:
Changed to another algorithm implementation.